File: 2024/11/main.go
package main

import (
	"adventofcode/u"
	"fmt"
)

func main() {
	nums := u.Nums(u.ReadFileStr("input.txt"), " ")

	res := 0
	blink := 75
	dp := make(map[nblink]int)
	for _, num := range nums {
		res += rec(num, blink, dp)
	}
	fmt.Println(res)
}

type nblink struct {
	n, blink int
}

func rec(n int, blink int, dp map[nblink]int) (res int) {
	if blink == 0 {
		return 1
	}
	if r, ok := dp[nblink{n, blink}]; ok {
		return r
	}
	defer func() { dp[nblink{n, blink}] = res }()

	if n == 0 {
		return rec(1, blink-1, dp)
	}

	str := fmt.Sprintf("%d", n)
	if len(str)%2 == 0 {
		return rec(u.Num(str[:len(str)/2]), blink-1, dp) + rec(u.Num(str[len(str)/2:]), blink-1, dp)
	}

	return rec(n*2024, blink-1, dp)
}

File: 2024/11/main.py
from functools import cache

@cache
def rec(n, b):
  if b == 0:
    return 1
  if n == 0:
    return rec(1, b - 1)
  n_str = str(n)
  if len(n_str) % 2 == 0:
    half = len(n_str) // 2
    return rec(int(n_str[:half]), b - 1) + rec(int(n_str[half:]), b - 1)
  return rec(n * 2024, b - 1)

with open("input.txt") as file:
  nums = [int(num_str) for num_str in file.readline().split()]
  res = sum(rec(num, 75) for num in nums)
  print(res)
